xarray/core/resample.py

```diff
@@ -0,0 +1,353 @@
+    specified dimension
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        self._dim = kwargs.pop('dim', None)
+        self._resample_dim = kwargs.pop('resample_dim', None)
+
+        if self._dim == self._resample_dim:
+            raise ValueError("Proxy resampling dimension ('{}') "
+                             "cannot have the same name as actual dimension "
+                             "('{}')! ".format(self._resample_dim, self._dim))
+        super(DataArrayResample, self).__init__(*args, **kwargs)
+
+    def apply(self, func, shortcut=False, **kwargs):
+        """Apply a function over each array in the group and concatenate them
+        together into a new array.
+
+        `func` is called like `func(ar, *args, **kwargs)` for each array `ar`
+        in this group.
+
+        Apply uses heuristics (like `pandas.GroupBy.apply`) to figure out how
+        to stack together the array. The rule is:
+        1. If the dimension along which the group coordinate is defined is
+           still in the first grouped array after applying `func`, then stack
+           over this dimension.
+        2. Otherwise, stack over the new dimension given by name of this
+           grouping (the argument to the `groupby` function).
+
+        Parameters
+        ----------
+        func : function
+            Callable to apply to each array.
+        shortcut : bool, optional
+            Whether or not to shortcut evaluation under the assumptions that:
+            (1) The action of `func` does not depend on any of the array
+                metadata (attributes or coordinates) but only on the data and
+                dimensions.
+            (2) The action of `func` creates arrays with homogeneous metadata,
+                that is, with the same dimensions and attributes.
+            If these conditions are satisfied `shortcut` provides significant
+            speedup. This should be the case for many common groupby operations
+            (e.g., applying numpy ufuncs).
+        **kwargs
+            Used to call `func(ar, **kwargs)` for each array `ar`.
+
+        Returns
+        -------
+        applied : DataArray or DataArray
+            The result of splitting, applying and combining this array.
+        """
+        combined = super(DataArrayResample, self).apply(
+            func, shortcut=shortcut, **kwargs)
+
+        # If the aggregation function didn't drop the original resampling
+        # dimension, then we need to do so before we can rename the proxy
+        # dimension we used.
+        if self._dim in combined:
+            combined = combined.drop(self._dim)
+
+        if self._resample_dim in combined.dims:
+            combined = combined.rename({self._resample_dim: self._dim})
+
+        return combined
+
+    def _interpolate(self, kind='linear'):
+        """Apply scipy.interpolate.interp1d along resampling dimension."""
+        from .dataarray import DataArray
+
```

Comment:
let's import `interp1d` here, instead of at the module level. `scipy` is still a optional dependency.

Reply:
Gotcha - it needed to be imported under `DatasetResample`, too.